# 4.5: Data Consistency Checks


## Contents

### 01. Importing libraries
### 02. Importing data
### 03. Data concistency checks

## 01. Importing libraries

In [9]:
# Import libraries
import pandas as pd
import numpy as np
import os

## 02. Importing data 

In [10]:
path = r'C:\Users\Sulo\Desktop\Careerfoundry\11.09.2023 Instacart Basket Analysis'

In [11]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [12]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

## 03. Data concistency checks

In [13]:
df_ords.describe()

Unnamed: 0      order_id       user_id  orders_amount  \
count  3.421083e+06  3.421083e+06  3.421083e+06   3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05   1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04   1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00   1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04   5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05   1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05   2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05   1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

In [14]:
# Create a dataframe
df_test = pd.DataFrame()

In [15]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [16]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [17]:
# Check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [18]:
# Change data type to string 
df_test['mix'] = df_test['mix'].astype('str')

In [19]:
# Check if there are null values and sum them up
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [20]:
# Create a dataframe that shows the null values 
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [21]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [22]:
df_prods.shape

(49693, 5)

In [23]:
# Create a dataframe that shows all the values without missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [24]:
df_prods_clean.shape

(49677, 5)

In [25]:
# Look for full duplicates in my dataframe
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [26]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [27]:
df_prods_clean.shape

(49677, 5)

In [28]:
# Create a new dataframe with clean data and no duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [29]:
df_prods_clean_no_dups.shape

(49672, 5)

In [30]:
# Export to csv
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

In [32]:
df_ords.describe()

Unnamed: 0      order_id       user_id  orders_amount  \
count  3.421083e+06  3.421083e+06  3.421083e+06   3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05   1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04   1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00   1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04   5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05   1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05   2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05   1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

## 2) I have a unnamed column which is bothersome and I don't know what this was for. Everything looks good but just keep in mind that with orders_amount there is a huge jump from 23 at 75% to max 100. So you would take the median over the mean if you wanted to compute missing values.

In [34]:
# Check for mixed types
for col in df_ords.columns.tolist():
  weirdo = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weirdo]) > 0:
    print (col)

In [35]:
# Check if there are null values and sum them up
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
orders_amount                  0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

## 5) There is no need the worry about the missing values above because it's reffering to 206209 values that just mean 0 days since prior order. There could be actual missing values in between there but most likely there isn't but a check is needed

In [36]:
# Create a dataframe that shows the null values 
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [37]:
df_ords_nan

Unnamed: 0  order_id  user_id  orders_amount  orders_day_of_week  \
0                 0   2539329        1              1                   2   
11               11   2168274        2              1                   2   
26               26   1374495        3              1                   1   
39               39   3343014        4              1                   6   
45               45   2717275        5              1                   3   
...             ...       ...      ...            ...                 ...   
3420930     3420930    969311   206205              1                   4   
3420934     3420934   3189322   206206              1                   3   
3421002     3421002   2166133   206207              1                   6   
3421019     3421019   2227043   206208              1                   1   
3421069     3421069   3154581   206209              1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

In [38]:
df_ords.median()

Unnamed: 0                1710541.0
order_id                  1710542.0
user_id                    102689.0
orders_amount                  11.0
orders_day_of_week              3.0
order_hour_of_day              13.0
days_since_prior_order          7.0
dtype: float64

In [42]:
# Compute the missing values with the median value
df_ords['days_since_prior_order'].fillna(7.0 , inplace=True)

In [43]:
df_ords

Unnamed: 0  order_id  user_id  orders_amount  orders_day_of_week  \
0                 0   2539329        1              1                   2   
1                 1   2398795        1              2                   3   
2                 2    473747        1              3                   3   
3                 3   2254736        1              4                   4   
4                 4    431534        1              5                   4   
...             ...       ...      ...            ...                 ...   
3421078     3421078   2266710   206209             10                   5   
3421079     3421079   1854736   206209             11                   4   
3421080     3421080    626363   206209             12                   1   
3421081     3421081   2977660   206209             13                   1   
3421082     3421082    272231   206209             14                   6   

         order_hour_of_day  days_since_prior_order  
0                        8                     7.0  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 7 columns]

## 6) I used the compute method because we had numbers to play with. The choice to take the median instead of the mean is because if you look at the 75% value we had 15 days instead of 22.5, that would have skewed the result and that's why I chose the median to have more consistency

In [44]:
# Look for full duplicates in my dataframe
df_dups_ords = df_ords[df_ords.duplicated()]

In [45]:
df_dups_ords

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, orders_amount, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

## 7) I've found no duplicates in the df_ords dataframe and therefore nothing needed to be done here and the next task

In [46]:
# Export to csv
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'ords_checked.csv'))